xbmc/guilib/GUIFontTTF.cpp

```diff
@@ -615,7 +627,93 @@ unsigned int CGUIFontTTF::GetTextureLineHeight() const
   return m_cellHeight + spacing_between_characters_in_texture;
 }
 
-CGUIFontTTF::Character* CGUIFontTTF::GetCharacter(character_t chr)
+vecGlyphInfo CGUIFontTTF::GetHarfbuzzShapedGlyphs(const vecText& text)
```

Comment:
Any chance this can be written without so many loops? You seem to iterate twice over the same values

Reply:
Let me explain the logic.

In order harfbuzz shaping to work, it need to know which script (like `HB_SCRIPT_LATIN` ,`HB_SCRIPT_MALAYALAM`, `HB_SCRIPT_ARABIC` etc ) the text is in advance. So the first loop determines the script of each charactors in text. In most of the cases , a single line of text contains only one script, but we need to make sure if there are multiscript texts.


The second iteration checks if there any `HB_SCRIPT_COMMON` or `HB_SCRIPT_INHERITED` ( these are for punctuation etc) in the script array we got from first loop and replaces it with script of it preceding letters.

The third loop is determing how many subtexts of different scripts are there in the text ( most cases it will be 1) and populates the `runs` vector

Fourth loop is doing the actual shaping based on the number of `runs`. If a text contains only one script ( in most of the cases) it will be run at once

Reply:
The loop inside the fourth loop adds character one by one to harfbuzz buffer ( `hb_buffer_add()`). We can add all at once( if they are of same script) with functions like `hb_buffer_add_utf8()` , `hb_buffer_add_utf16()` etc, but only  `hb_buffer_add()` lets us to provide a`cluster` field (  cluster refer to the index of the character in the input text so that we can determine which glyph of the output glyph info array correspond to the input text. This is important as the style/color information is stored in HIWORD of each  characters in kodi's `vecText`  and we need it later when rendering the character)

Reply:
I'm not a native English speaker, so please ask me if you didn't understand the logic I said. :)

Reply:
And the logic is same form vlc's harfbuzz code. But they don't need the `cluster` field so they add the whole string at oncento harfbuzz buffer. But for us `hb_glyph_info_t.cluster` is needed for a later stage.